src/uct/tcp/tcp.h

```diff
@@ -27,18 +33,30 @@ typedef struct uct_tcp_am_hdr {
 
 
 /**
- * TCP endpoint
+ * TCP endpoint communication context
  */
-typedef struct uct_tcp_ep {
-    uct_base_ep_t                 super;
-    int                           fd;        /* Socket file descriptor */
-    uint32_t                      events;    /* Current notifications */
-    ucs_queue_head_t              pending_q; /* Pending operations */
+typedef struct uct_tcp_ep_ctx {
     void                          *buf;      /* Partial send/recv data */
     size_t                        length;    /* How much data in the buffer */
     size_t                        offset;    /* Next offset to send/recv */
+} uct_tcp_ep_ctx_t;
+
+
+/**
+ * TCP endpoint
+ */
+struct uct_tcp_ep {
+    uct_base_ep_t                 super;
+    int                           fd;          /* Socket file descriptor */
+    uint32_t                      events;      /* Current notifications */
+    uct_tcp_ep_ctx_t              tx;          /* TX resources */
```

Comment:
minor: you can combine tx and rx fields under `tx` and `rx` substructs, like we do for other transports.

Reply:
do you mean anonymous structures inside uct_tcp_ep_t?
i.e.:
```
struct uct_tcp_ep {
...
    struct {
        void *buf;
        size_t lenght;
        size_t offset;
    } tx;
    struct {
        void *buf;
        size_t lenght;
        size_t offset;
    } rx;
...
};
```

Reply:
No, more like:
```
struct {
    uct_tcp_ep_ctx_t              ctx;
    uct_tcp_ep_progress_t     progress;
    ucs_queue_head_t           pending_q;
} tx;
struct {
    uct_tcp_ep_ctx_t              ctx;
    uct_tcp_ep_progress_t     progress;
 } rx;
 ```

BTW, maybe it is worth to add progress function to the ep context?

Reply:
> BTW, maybe it is worth to add progress function to the ep context instead?

Good idea!
will follow your proposal. Thank you!